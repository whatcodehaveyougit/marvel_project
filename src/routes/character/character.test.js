import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Character from './character';
import characterData from './testData/character-object.json';
import comicData from './testData/comic-object.json';
import { useParams } from 'react-router-dom';
// import { useSelector } from 'react-redux';
import * as reactRedux from 'react-redux'

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'), // Preserve all original exports
  useSelector: jest.fn(),
  useDispatch: jest.fn(() => jest.fn()), // Mock useDispatch
}));

jest.mock('react-router-dom', () => ({
  useParams: jest.fn(),
}));

jest.mock('../../components/spinner/spinner.component', () => jest.fn(() => 'Spinner'));

const mockStore = configureStore([]);

// const initialState = {
//   characters: [ characterData ], // TO FIX: Ensure characters data is initially an empty array
//   isLoading: false,
//   error: null,
//   comics: [],
// };

describe('Character', () => {
  // const useSelectorMock = jest.spyOn(reactRedux, 'useSelector')

  beforeEach(() => {
    // Set up useSelector mock to return initialState.characters
    // useSelectorMock.mockReturnValue(initialState);
    useParams.mockReturnValue({ characterid: 1011334 });
  });

  afterEach(() => {
    // Clear mock implementation after each test
    // useSelector.mockClear();
  });

  // it('Component renders correctly', () => {
  //   const store = mockStore(initialState);
  //   const { asFragment } = render(
  //     <Provider store={store}>
  //       <Character />
  //     </Provider>
  //   );
  //   expect(asFragment()).toMatchSnapshot();
  // });

  it('Spinner is rendered when there is no comic book data but the data is loading', () => {
    const initialState = {
      characters: [ characterData ], // TO FIX: Ensure characters data is initially an empty array
      isLoading: false,
      error: null,
      character: {
        comics: [],
        data: null,
        isLoading: true,
        error: null
      }
    };
    const store = mockStore(initialState);
    jest.spyOn(reactRedux, 'useSelector').mockImplementation(callback => callback(store.getState()));;
    const view = render(
      <Provider store={store}>
        <Character />
      </Provider>
    );
    expect(view).toMatchSnapshot();
  });

  it('Accordion of comics is rendered when data has loaded in', () => {
    const initialState1 = {
      characters: [ characterData ], // TO FIX: Ensure characters data is initially an empty array
      isLoading: false,
      error: null,
      character: {
        comics: [ comicData ],
        data: { name: 'THe hulk'},
        isLoading: false,
        error: null
      }
    };
    const store = mockStore(initialState1);
    jest.spyOn(reactRedux, 'useSelector').mockImplementation(callback => callback(store.getState()));
    const view = render(
      <Provider store={store}>
        <Character />
      </Provider>
    );
    expect(view).toMatchSnapshot();
  });


});

