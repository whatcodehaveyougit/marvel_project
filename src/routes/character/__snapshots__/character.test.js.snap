// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="character-page bg-gray-100"
  >
    <div
      class="thumbnail-image bg-cover bg-center"
      style="background-image: url();"
    >
      <div
        class="character-information p-8"
      >
        <div
          class="character-page-headings"
        >
          <h3
            class="text-3xl text-center"
          />
          <h5
            class="text-lg text-center"
          >
            List of comics for this character:
          </h5>
        </div>
        <div>
          <div
            class="fixed inset-0 flex items-center justify-center bg-white bg-opacity-90"
          >
            <div
              class="w-16 h-16 border-8 border-blue-500 border-t-transparent border-r-transparent rounded-full animate-spin"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Character renders correctly 1`] = `
<DocumentFragment>
  <div
    class="character-page bg-gray-100"
  >
    <div
      class="thumbnail-image bg-cover bg-center"
      style="background-image: url(http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg);"
    >
      <div
        class="character-information p-8"
      >
        <div
          class="character-page-headings"
        >
          <h3
            class="text-3xl text-center"
          >
            3-D Man
          </h3>
          <h5
            class="text-lg text-center"
          >
            List of comics for this character:
          </h5>
        </div>
        <div />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
